zebra-state/src/service.rs

```diff
@@ -1247,21 +1247,39 @@ impl Service<ReadRequest> for ReadStateService {
             }
 
             // For the get_raw_transaction RPC and the StateService.
-            ReadRequest::Transaction(hash) => {
+            ReadRequest::Transaction {
+                hash,
+                should_return_confirmations,
+            } => {
                 let state = self.clone();
 
                 tokio::task::spawn_blocking(move || {
                     span.in_scope(move || {
-                        let transaction_and_height = state
-                            .non_finalized_state_receiver
-                            .with_watch_data(|non_finalized_state| {
-                                read::transaction(non_finalized_state.best_chain(), &state.db, hash)
-                            });
+                        let latest_best_chain = state.latest_best_chain();
+
+                        let transaction_and_height =
+                            read::transaction(latest_best_chain.as_ref(), &state.db, hash);
+
+                        let confirmations = should_return_confirmations
+                            .then(|| {
+                                transaction_and_height.as_ref().and_then(|(_tx, height)| {
+                                    latest_best_chain
+                                        .map(|latest_best_chain| {
+                                            latest_best_chain.non_finalized_tip_height()
+                                        })
+                                        .or_else(|| state.db.finalized_tip_height())
```

Comment:
What happens if a block is pushed to the finalized state after we clone the `latest_best_chain`?
(I think that both the lower and higher values are ok, but we need a correctness comment to explain why, including a justification that the subtraction won't panic.)

Optional: this code could also get moved to the `read::block` module, along with the rest of the complex queries with correctness comments.

Reply:
> What happens if a block is pushed to the finalized state after we clone the `latest_best_chain`? (I think that both the lower and higher values are valid, but we need a correctness comment to explain why, including a justification that the subtraction won't panic.)

It would read the more current finalized tip height if the non-finalized chain is empty, or a potentially outdated non-finalized tip height that's current at the time that it's cloned. I left the same comment as the `read::depth()` function in [`6cb2823`](https://github.com/ZcashFoundation/zebra/pull/6287/commits/6cb2823c015f866aecf139902ce6b192ced99a10)
 
> Optional: this code could also get moved to the `read::block` module, along with the rest of the complex queries with correctness comments.

Moved to `read::block::mined_transaction()` function in [`6cb2823`](https://github.com/ZcashFoundation/zebra/pull/6287/commits/6cb2823c015f866aecf139902ce6b192ced99a10)